neqo-transport/src/frame.rs

```diff
@@ -18,68 +18,152 @@ use crate::{
-const FRAME_TYPE_STREAM_MAX: FrameType = 0xf;
-pub const FRAME_TYPE_MAX_DATA: FrameType = 0x10;
-pub const FRAME_TYPE_MAX_STREAM_DATA: FrameType = 0x11;
-pub const FRAME_TYPE_MAX_STREAMS_BIDI: FrameType = 0x12;
-pub const FRAME_TYPE_MAX_STREAMS_UNIDI: FrameType = 0x13;
-pub const FRAME_TYPE_DATA_BLOCKED: FrameType = 0x14;
-pub const FRAME_TYPE_STREAM_DATA_BLOCKED: FrameType = 0x15;
-pub const FRAME_TYPE_STREAMS_BLOCKED_BIDI: FrameType = 0x16;
-pub const FRAME_TYPE_STREAMS_BLOCKED_UNIDI: FrameType = 0x17;
-pub const FRAME_TYPE_NEW_CONNECTION_ID: FrameType = 0x18;
-pub const FRAME_TYPE_RETIRE_CONNECTION_ID: FrameType = 0x19;
-pub const FRAME_TYPE_PATH_CHALLENGE: FrameType = 0x1a;
-pub const FRAME_TYPE_PATH_RESPONSE: FrameType = 0x1b;
-pub const FRAME_TYPE_CONNECTION_CLOSE_TRANSPORT: FrameType = 0x1c;
-pub const FRAME_TYPE_CONNECTION_CLOSE_APPLICATION: FrameType = 0x1d;
-pub const FRAME_TYPE_HANDSHAKE_DONE: FrameType = 0x1e;
-// draft-ietf-quic-ack-delay
-pub const FRAME_TYPE_ACK_FREQUENCY: FrameType = 0xaf;
-// draft-ietf-quic-datagram
-pub const FRAME_TYPE_DATAGRAM: FrameType = 0x30;
-pub const FRAME_TYPE_DATAGRAM_WITH_LEN: FrameType = 0x31;
-const DATAGRAM_FRAME_BIT_LEN: u64 = 0x01;
-
-const STREAM_FRAME_BIT_FIN: u64 = 0x01;
-const STREAM_FRAME_BIT_LEN: u64 = 0x02;
-const STREAM_FRAME_BIT_OFF: u64 = 0x04;
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum FrameType {
+    Padding = 0x0,
+    Ping = 0x1,
+    Ack = 0x2,
+    AckEcn = 0x3,
+    ResetStream = 0x4,
+    StopSending = 0x5,
+    Crypto = 0x6,
+    NewToken = 0x7,
+    Stream = 0x08, // + 0b000
+    StreamWithFin = 0x08 + 0b001,
+    StreamWithLen = 0x08 + 0b010,
+    StreamWithLenFin = 0x08 + 0b011,
+    StreamWithOff = 0x08 + 0b100,
+    StreamWithOffFin = 0x08 + 0b101,
+    StreamWithOffLen = 0x08 + 0b110,
+    StreamWithOffLenFin = 0x08 + 0b111,
+    MaxData = 0x10,
+    MaxStreamData = 0x11,
+    MaxStreamsBiDi = 0x12,
+    MaxStreamsUniDi = 0x13,
+    DataBlocked = 0x14,
+    StreamDataBlocked = 0x15,
+    StreamsBlockedBiDi = 0x16,
+    StreamsBlockedUniDi = 0x17,
+    NewConnectionId = 0x18,
+    RetireConnectionId = 0x19,
+    PathChallenge = 0x1a,
+    PathResponse = 0x1b,
+    ConnectionCloseTransport = 0x1c,
+    ConnectionCloseApplication = 0x1d,
+    HandshakeDone = 0x1e,
+    // draft-ietf-quic-ack-delay
+    AckFrequency = 0xaf,
+    // draft-ietf-quic-datagram
+    Datagram = 0x30,
+    DatagramWithLen = 0x31,
+}
 
-#[derive(PartialEq, Eq, Debug, PartialOrd, Ord, Clone, Copy)]
-pub enum CloseError {
-    Transport(TransportError),
-    Application(AppError),
+impl From<FrameType> for u64 {
+    fn from(val: FrameType) -> Self {
+        val as Self
+    }
 }
 
-impl CloseError {
-    const fn frame_type_bit(self) -> u64 {
-        match self {
-            Self::Transport(_) => 0,
-            Self::Application(_) => 1,
+impl From<FrameType> for u8 {
+    fn from(val: FrameType) -> Self {
+        val as Self
+    }
+}
+
+impl TryFrom<u64> for FrameType {
```

Comment:
It seems to me like this is ripe for a derive-like macro.  I wonder if there are any of those floating around?

Reply:
Oh yeah, there are a ton of useful things in `strum`.